cvc5/Init.lean

```diff
@@ -0,0 +1,193 @@
+        ident.getId.toString
+    pref ++ "_" ++ id |> Syntax.mkStrLit
+  let mods ←
+    match mods with
+    | `(Parser.Command.declModifiersT|
+      $[$doc:docComment]? $[@[ $[ $attrs ],* ]]? $[$vis]? $[$isNc]? $[$isUnsafe]? $[$opt]?
+    ) => do
+      let ext ← `(Parser.Term.attrInstance| extern $externName:str)
+      let attrs := attrs.getD #[] |>.push ext
+      `(Parser.Command.declModifiersT|
+        $[$doc:docComment]? @[ $[$attrs],* ] $[$vis]? $[$isNc]? $[$isUnsafe]? $[$opt]?
+      )
+    | _ => throwUnsupportedSyntax
+  let mainDef ←`(
+    set_option linter.unusedVariables false in
+    $(⟨mods⟩):declModifiers
+    opaque $ident:declId $identSig
+  )
+  Command.elabCommand mainDef
+
+  let define doc? id sig? (body : Syntax.Term) : CommandElabM _ := do
+    if let some doc := doc? then
+      `(command|
+        $doc:docComment
+        def $id:declId $sig?:optDeclSig := $body
+      )
+    else
+      `(command|
+        @[inherit_doc $ident]
+        def $id:declId $sig?:optDeclSig := $body
+      )
+
+  if let (some autoDoc?, some autoId) := (autoDoc, autoId) then
+    let env ← getEnv
+    let id := ident.getId.toString
+    let name := Lean.Name.mkSimple id
+    let ns ← Elab.liftMacroM Macro.getCurrNamespace
+    let name := ns.append name
+    let arity ←
+      if let some (.opaqueInfo i) := env.find? name then
+        pure (forallArity i.type)
+      else
+        throwError s!"failed to retrieve arity of (opaque) function `{ident}`"
+
+    let mut args := Array.empty
+    for i in [0:arity] do
+      let arg := Lean.Name.mkSimple s!"v{i}" |> Lean.mkIdent
+      args := args.push arg
+    let funCall : TSyntax `term ← `(term| ($ident $[ $args ]* ))
+
+    for (autoDoc?, autoId) in autoDoc?.zip autoId do
+      let id : String := autoId.getId.toString
+      let body ←
+        if id.endsWith "!" then
+          let unwrapName := Lean.Name.mkStr3 "cvc5" "Error" "unwrap!" |> Lean.mkIdent
+          `(fun $[$args]* => $funCall |> $unwrapName)
+        else if id.endsWith "?" then
+          `(fun $[$args]* => $funCall |> Except.toOption)
+        else
+          throwError s!"unexpected auto function name `{id}`: expected `<ident>!` or `<ident>?`"
+      Command.elabCommand
+        (← define autoDoc? autoId (← `(optDeclSig|)) body)
+
+  if let
+    (some subDoc?, some subId, some subSig, some subDef)
+    := (subDoc, subId, subSig, subDef)
+  then
+    let all := subDoc?.zip subId |>.zip subSig |>.zip subDef
+    for (((subDoc?, subId), subSig), subDef) in all do
+      Command.elabCommand
+        (← define subDoc? subId subSig subDef)
+| _ => throwUnsupportedSyntax
+
+/-- Defines similar functions realized by `extern`.
+
+```
+extern! "prefix"
+  /-- Create a Boolean constant.
+
+  - `b`: The Boolean constant.
+
+  Will create an opaque definition with `[@extern extStr]` where
+  `extStr = "prefix" ++ "_" ++ "myFunction"`.
+  -/
+  def myFunction : Term → Except Error Op
+  with
+    endsWithBang!
+    endWithQuestion?
```

Comment:
Could you also add macros like `extern_def`, `extern_def!`, and `extern_def?` that handle the simple cases for a single function? I don't foresee many use cases of special handling.

Reply:
To generate `!` and `?` functions it's easier to already know which `extern` function they are based on. We can directly build the custom definition since we already know the `extern` signature, and auto-add the `inherit_doc` when no documentation is provided.

Reply:
I understand your point. What I meant is that `extern_def!` should do what `extern_def` does in addition to boilerplate code for a panic version of the function. Similarly, `extern_def?` should be `extern_def` and boilerplate code for an option version. For both, we can define a macro like `extern_def!?`.